Maven 4 (rc5) fails in maven-install-plugin due to BUILD_CONSUMER disabling profile activation for locally-resolved parent POMs

### Affected version

4.0.0-rc5

### Bug description

Note that I asked Claude Code to help with the analysis and the fix.

This is a different bug from #11767, discovered while testing the fixes from that ticket. The #11767 fix (PR #11768) correctly passes repositories and profiles to the consumer POM builder, but a separate code path still causes failures when parent POMs define properties inside profiles with property-based activation.

 ```
 $ mvn clean install
[...]
[INFO] --- install:3.1.3:install (default-install) @ xxx ---
[INFO] --------------------------------------------------------------------------------------------------------------------------
 [INFO] BUILD FAILURE
 [INFO] --------------------------------------------------------------------------------------------------------------------------
 [...]
 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-install-plugin:3.1.3:install (default-install) on project xxx: Execution default-install of goal org.apache.maven.plugins:maven-install-plugin:3.1.3:install failed: Invalid Version Range Request: org.springframework:spring-framework-bom:pom:${spring.version} [...]
```
The property ${spring.version} is defined in a great-grandparent POM (lineup) inside a profile with property-based activation:
```
<!-- lineup pom.xml -->
  <profiles>
      <profile>
          <id>spring-old</id>
          <activation>
              <property>
                  <name>!explicitelyDeactivated</name>
              </property>
          </activation>
          <properties>
              <spring.version>5.3.39</spring.version>
          </properties>
      </profile>
  </profiles>
```

A grandparent POM uses this property in a BOM import:
```
  <dependencyManagement>
      <dependencies>
          <dependency>
              <groupId>org.springframework</groupId>
              <artifactId>spring-framework-bom</artifactId>
              <version>${spring.version}</version>
              <type>pom</type>
              <scope>import</scope>
          </dependency>
      </dependencies>
  </dependencyManagement>
```

During BUILD_PROJECT, profiles are properly activated and the property resolves to 5.3.39. During BUILD_CONSUMER (install phase), the property is null and the BOM import fails with Invalid Version Range Request.

**Root cause**
The bug is in DefaultModelBuilder.readParentLocally() (line ~1151 in DefaultModelBuilder.java).
When the consumer POM is built (BUILD_CONSUMER), readParentLocally() finds the parent POM via resolveReactorModel() (because the parent was loaded into mappedSources during the earlier BUILD_PROJECT phase). It then calls:
```
  // Line ~1151 - DefaultModelBuilder.java - readParentLocally()
  ModelBuilderSessionState derived = derive(candidateSource);
```
derive(candidateSource) calls derive(ModelBuilderRequest.build(request, source)) which preserves the BUILD_CONSUMER request type. Since isBuildRequestWithActivation() returns false for BUILD_CONSUMER, POM profile activation is skipped for all parent POMs in the chain:

```
  boolean isBuildRequestWithActivation() {
      return request.getRequestType() != ModelBuilderRequest.RequestType.BUILD_CONSUMER;
  }
```
In contrast, resolveAndReadParentExternally() (used when the parent is NOT in the reactor) explicitly creates a CONSUMER_PARENT request:
```
  // Line ~1284 - DefaultModelBuilder.java - resolveAndReadParentExternally()
  ModelBuilderRequest lenientRequest = ModelBuilderRequest.builder(request)
          .requestType(ModelBuilderRequest.RequestType.CONSUMER_PARENT)
          .source(modelSource)
          .build();
```
CONSUMER_PARENT allows profile activation, so parent POMs resolved externally work correctly. Only parent POMs resolved locally (via reactor/mappedSources) are affected.

**Fix**
In readParentLocally(), when the current request is BUILD_CONSUMER, derive the parent session with CONSUMER_PARENT type — consistent with how resolveAndReadParentExternally() handles it:

```
  ModelBuilderSessionState derived;
  if (request.getRequestType() == ModelBuilderRequest.RequestType.BUILD_CONSUMER) {
      ModelBuilderRequest parentRequest = ModelBuilderRequest.builder(request)
              .requestType(ModelBuilderRequest.RequestType.CONSUMER_PARENT)
              .source(candidateSource)
              .build();
      derived = derive(parentRequest);
  } else {
      derived = derive(candidateSource);
  }
```
Claude Code created the following unit test (which fails without the fix):
Test: DefaultModelBuilderTest.testBuildConsumerResolvesParentProfileProperties (maven-impl)
  - Builds a parent POM with BUILD_PROJECT to populate the reactor (mappedSources) — the parent defines managed.version=1.2.3 inside a profile with property-based activation (!skipDefaultVersions)
  - Builds the child POM with BUILD_CONSUMER on the same session — the parent is found via resolveReactorModel() in readParentLocally()
  - Asserts the effective model has managed.version=1.2.3 (from the parent's profile)
  - Asserts the managed dependency version is interpolated to 1.2.3, not ${managed.version}
Without the fix: `AssertionFailedError: expected: <1.2.3> but was: <null>`

I was also able to confirm that when implementing the fix on top of 4.0-rc5, my mvn clean install passed.

I am now preparing a Pull Request.

--- a/impl/maven-impl/src/main/java/org/apache/maven/impl/model/DefaultModelBuilder.java
+++ b/impl/maven-impl/src/main/java/org/apache/maven/impl/model/DefaultModelBuilder.java
@@ -1160,7 +1160,13 @@ public class DefaultModelBuilder implements ModelBuilder {
             }
 
             try {
-                ModelBuilderSessionState derived = derive(candidateSource);
+                ModelBuilderSessionState derived = derive(
+                        request.getRequestType() == ModelBuilderRequest.RequestType.BUILD_CONSUMER
+                                ? ModelBuilderRequest.builder(request)
+                                        .requestType(ModelBuilderRequest.RequestType.CONSUMER_PARENT)
+                                        .source(candidateSource)
+                                        .build()
+                                : ModelBuilderRequest.build(request, candidateSource));
 
                 // Check GA match BEFORE readAsParentModel() which recursively resolves
                 // the candidate's parent chain and can trigger false cycle detection (GH-12074).
@@ -1174,7 +1180,6 @@ public class DefaultModelBuilder implements ModelBuilder {
                     mismatchRelativePathAndGA(childModel, parent, fileGroupId, fileArtifactId);
                     return null;
                 }
-
                 Model candidateModel = derived.readAsParentModel(profileActivationContext, parentChain);
                 // Add profiles from parent, preserving model ID tracking
                 for (Map.Entry<String, List<Profile>> entry :

--- a/impl/maven-impl/src/test/java/org/apache/maven/impl/model/DefaultModelBuilderTest.java
+++ b/impl/maven-impl/src/test/java/org/apache/maven/impl/model/DefaultModelBuilderTest.java
@@ -416,7 +416,7 @@ class DefaultModelBuilderTest {
         assertEquals(
                 "1.2.3",
                 managedDep.getVersion(),
-                "Managed dependency version should be interpolated, not ${managed.version}")
+                "Managed dependency version should be interpolated, not ${managed.version}");
     }
 
     private Path getPom(String name) {

--- a/impl/maven-impl/src/test/java/org/apache/maven/impl/model/DefaultModelBuilderTest.java
+++ b/impl/maven-impl/src/test/java/org/apache/maven/impl/model/DefaultModelBuilderTest.java
@@ -38,6 +38,7 @@ import org.apache.maven.impl.standalone.ApiRunner;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertNull;
@@ -370,6 +371,52 @@ class DefaultModelBuilderTest {
         assertTrue(
                 derived.getExternalRepositories().stream().anyMatch(r -> "custom-repo".equals(r.getId())),
                 "Derived session externalRepositories should include the custom repo from the request");
+    }
+
+    /**
+     * Verifies that BUILD_CONSUMER resolves properties defined in parent POM profiles
+     * when the parent is found via reactor model resolution (mappedSources).
+     */
+    @Test
+    public void testBuildConsumerResolvesParentProfileProperties() {
+        Path parentPom = getPom("consumer-profile-property-parent");
+        Path childPom = getPom("consumer-profile-property-child");
+
+        ModelBuilder.ModelBuilderSession mbs = builder.newSession();
+
+        mbs.build(ModelBuilderRequest.builder()
+                .session(session)
+                .requestType(ModelBuilderRequest.RequestType.BUILD_PROJECT)
+                .source(Sources.buildSource(parentPom))
+                .build());
+
+        ModelBuilderResult consumerResult = assertDoesNotThrow(
+                () -> mbs.build(ModelBuilderRequest.builder()
+                        .session(session)
+                        .requestType(ModelBuilderRequest.RequestType.BUILD_CONSUMER)
+                        .source(Sources.buildSource(childPom))
+                        .build()),
+                "BUILD_CONSUMER should not fail when parent defines properties in profiles");
+
+        assertNotNull(consumerResult);
+        Model effectiveModel = consumerResult.getEffectiveModel();
+        assertNotNull(effectiveModel);
+
+        assertEquals(
+                "1.2.3",
+                effectiveModel.getProperties().get("managed.version"),
+                "Property from parent's profile should be resolved in BUILD_CONSUMER effective model");
+
+        assertNotNull(effectiveModel.getDependencyManagement());
+        Dependency managedDep = effectiveModel.getDependencyManagement().getDependencies().stream()
+                .filter(d -> "managed-lib".equals(d.getArtifactId()))
+                .findFirst()
+                .orElse(null);
+        assertNotNull(managedDep, "Managed dependency from parent should be inherited");
+        assertEquals(
+                "1.2.3",
+                managedDep.getVersion(),
+                "Managed dependency version should be interpolated, not ${managed.version}")
     }
 
     private Path getPom(String name) {

--- a/impl/maven-impl/src/test/resources/poms/factory/consumer-profile-property-child.xml
+++ b/impl/maven-impl/src/test/resources/poms/factory/consumer-profile-property-child.xml
@@ -1,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+ Licensed to the Apache Software Foundation (ASF) under one or more
+ contributor license agreements.  See the NOTICE file distributed with
+ this work for additional information regarding copyright ownership.
+ The ASF licenses this file to You under the Apache License, Version 2.0
+ (the "License"); you may not use this file except in compliance with
+ the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-->
+<project xmlns="http://maven.apache.org/POM/4.1.0">
+    <parent>
+        <groupId>org.apache.maven.tests</groupId>
+        <artifactId>consumer-profile-property-parent</artifactId>
+        <relativePath>consumer-profile-property-parent.xml</relativePath>
+    </parent>
+    <artifactId>consumer-profile-property-child</artifactId>
+    <version>1.0-SNAPSHOT</version>
+    <packaging>jar</packaging>
+</project>

--- a/impl/maven-impl/src/test/resources/poms/factory/consumer-profile-property-parent.xml
+++ b/impl/maven-impl/src/test/resources/poms/factory/consumer-profile-property-parent.xml
@@ -1,0 +1,52 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+ Licensed to the Apache Software Foundation (ASF) under one or more
+ contributor license agreements.  See the NOTICE file distributed with
+ this work for additional information regarding copyright ownership.
+ The ASF licenses this file to You under the Apache License, Version 2.0
+ (the "License"); you may not use this file except in compliance with
+ the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-->
+<project xmlns="http://maven.apache.org/POM/4.1.0">
+    <groupId>org.apache.maven.tests</groupId>
+    <artifactId>consumer-profile-property-parent</artifactId>
+    <version>1.0-SNAPSHOT</version>
+    <packaging>pom</packaging>
+
+    <!--
+      The property "managed.version" is ONLY defined inside a profile
+      with property-based activation. This simulates the real-world
+      scenario where parent POMs define version properties in profiles.
+    -->
+    <profiles>
+        <profile>
+            <id>default-versions</id>
+            <activation>
+                <property>
+                    <name>!skipDefaultVersions</name>
+                </property>
+            </activation>
+            <properties>
+                <managed.version>1.2.3</managed.version>
+            </properties>
+        </profile>
+    </profiles>
+
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>org.example</groupId>
+                <artifactId>managed-lib</artifactId>
+                <version>${managed.version}</version>
+            </dependency>
+        </dependencies>
+    </dependencyManagement>
+</project>

--- a/its/core-it-suite/src/test/java/org/apache/maven/it/MavenITgh11798ConsumerPomProfileActivationTest.java
+++ b/its/core-it-suite/src/test/java/org/apache/maven/it/MavenITgh11798ConsumerPomProfileActivationTest.java
@@ -1,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.maven.it;
+
+import java.nio.file.Path;
+
+import org.junit.jupiter.api.Test;
+
+/**
+ * Verify that consumer POM generation preserves profile activation for locally-resolved parent POMs.
+ * When a parent POM defines properties inside a profile with property-based activation,
+ * those properties must be resolved during BUILD_CONSUMER so that BOM imports using
+ * those properties do not fail with "Invalid Version Range Request".
+ *
+ * @see <a href="https://github.com/apache/maven/issues/11798">GH-11798</a>
+ */
+class MavenITgh11798ConsumerPomProfileActivationTest extends AbstractMavenIntegrationTestCase {
+
+    @Test
+    void testConsumerPomResolvesParentProfileProperties() throws Exception {
+        Path basedir = extractResources("/gh-11798-consumer-pom-profile-activation").toPath();
+
+        Verifier verifier = newVerifier(basedir.toString());
+        verifier.addCliArgument("install");
+        verifier.execute();
+        verifier.verifyErrorFreeLog();
+    }
+}

--- a/its/core-it-suite/src/test/resources/gh-11798-consumer-pom-profile-activation/child/pom.xml
+++ b/its/core-it-suite/src/test/resources/gh-11798-consumer-pom-profile-activation/child/pom.xml
@@ -1,0 +1,43 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+<project xmlns="http://maven.apache.org/POM/4.1.0">
+  <modelVersion>4.1.0</modelVersion>
+
+  <parent>
+    <groupId>org.apache.maven.its.gh11798</groupId>
+    <artifactId>parent</artifactId>
+    <version>1.0</version>
+  </parent>
+
+  <artifactId>child</artifactId>
+  <packaging>jar</packaging>
+
+  <dependencyManagement>
+    <dependencies>
+      <dependency>
+        <groupId>org.apache.maven.its.gh11798</groupId>
+        <artifactId>test-bom</artifactId>
+        <version>${dep.version}</version>
+        <type>pom</type>
+        <scope>import</scope>
+      </dependency>
+    </dependencies>
+  </dependencyManagement>
+</project>

--- a/its/core-it-suite/src/test/resources/gh-11798-consumer-pom-profile-activation/pom.xml
+++ b/its/core-it-suite/src/test/resources/gh-11798-consumer-pom-profile-activation/pom.xml
@@ -1,0 +1,59 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+<project xmlns="http://maven.apache.org/POM/4.1.0" root="true">
+  <modelVersion>4.1.0</modelVersion>
+
+  <groupId>org.apache.maven.its.gh11798</groupId>
+  <artifactId>parent</artifactId>
+  <version>1.0</version>
+  <packaging>pom</packaging>
+
+  <modules>
+    <module>child</module>
+  </modules>
+
+  <repositories>
+    <repository>
+      <releases>
+        <enabled>true</enabled>
+        <checksumPolicy>ignore</checksumPolicy>
+      </releases>
+      <snapshots>
+        <enabled>false</enabled>
+      </snapshots>
+      <id>local-test-repo</id>
+      <url>file://${project.rootDirectory}/repo</url>
+    </repository>
+  </repositories>
+
+  <profiles>
+    <profile>
+      <id>default-versions</id>
+      <activation>
+        <property>
+          <name>!skipDefaultVersions</name>
+        </property>
+      </activation>
+      <properties>
+        <dep.version>1.0</dep.version>
+      </properties>
+    </profile>
+  </profiles>
+</project>

--- a/its/core-it-suite/src/test/resources/gh-11798-consumer-pom-profile-activation/repo/org/apache/maven/its/gh11798/test-bom/1.0/test-bom-1.0.pom
+++ b/its/core-it-suite/src/test/resources/gh-11798-consumer-pom-profile-activation/repo/org/apache/maven/its/gh11798/test-bom/1.0/test-bom-1.0.pom
@@ -1,0 +1,7 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0">
+    <modelVersion>4.0.0</modelVersion>
+    <groupId>org.apache.maven.its.gh11798</groupId>
+    <artifactId>test-bom</artifactId>
+    <version>1.0</version>
+    <packaging>pom</packaging>
+</project>

